src/Backends/PrefersColorSchemeSettings.vala

```diff
@@ -0,0 +1,120 @@
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+* Authored by: Marius Meisenzahl <mariusmeisenzahl@gmail.com>
+*/
+
+public class SettingsDaemon.Backends.PrefersColorSchemeSettings : GLib.Object {
+    public unowned PantheonShell.Pantheon.AccountsService accounts_service { get; construct; }
+
+    private GLib.Settings color_settings;
+    private double pos_lat = -1.0;
+    private double pos_long = -1.0;
+
+    public PrefersColorSchemeSettings (PantheonShell.Pantheon.AccountsService accounts_service) {
+        Object (accounts_service: accounts_service);
+    }
+
+    construct {
+        color_settings = new GLib.Settings ("io.elementary.settings-daemon.prefers-color-scheme");
+
+        get_location.begin ();
+
+        var time = new TimeoutSource (1000);
+
+        time.set_callback (() => {
+            var schedule = color_settings.get_string ("prefer-dark-schedule");
+
+            var now = new DateTime.now_local ();
+            double from, to;
+            if (schedule == "sunset-to-sunrise") {
+                double sunrise, sunset;
+
+                bool success = SettingsDaemon.Utils.SunriseSunsetCalculator.get_sunrise_and_sunset (now, pos_lat, pos_long, out sunrise, out sunset);
+
+                if (success) {
+                    from = sunset;
+                    to = sunrise;
+                }
+            } else if (schedule == "manual") {
+                from = color_settings.get_double ("prefer-dark-schedule-from");
+                to = color_settings.get_double ("prefer-dark-schedule-to");
+            } else {
+                return true;
+            }
+
+            var state = get_state (date_time_double (now), from, to);
+            var new_color_scheme = Granite.Settings.ColorScheme.NO_PREFERENCE;
+            if (state == State.IN) {
+                new_color_scheme = Granite.Settings.ColorScheme.DARK;
+            }
+
+            if (new_color_scheme == accounts_service.prefers_color_scheme) {
+                return true;
+            }
+
+            accounts_service.prefers_color_scheme = new_color_scheme;
+
+            return true;
+        });
+
+        time.attach (null);
+    }
+
+    private async void get_location () {
+        try {
+            var simple = yield new GClue.Simple (Build.PROJECT_NAME, GClue.AccuracyLevel.CITY, null);
+
+            simple.notify["location"].connect (() => {
+                on_location_updated (simple.location.latitude, simple.location.longitude);
+            });
+
+            on_location_updated (simple.location.latitude, simple.location.longitude);
+        } catch (Error e) {
+            warning ("Failed to connect to GeoClue2 service: %s", e.message);
+            return;
+        }
+    }
+
+    private void on_location_updated (double latitude, double longitude) {
+        pos_lat = latitude;
+        pos_long = longitude;
+    }
+
+    private enum State {
```

Comment:
I'm not sure we really need an enum at this stage, just a boolean of `is_in_time_window` or something would be fine.